include/xtypes/DynamicDataImpl.hpp

```diff
@@ -274,318 +274,424 @@ inline DynamicData DynamicData::operator - () const
-            return *this;
+            xtypes_assert(false, "operator--() isn't available for type '" << type_.name() <<  "'.");
     }
+}
 
+inline DynamicData& DynamicData::operator -- ()
+{
+    WritableDynamicDataRef wthis = this->ref();
+    wthis = DynamicData::operator--(1);
+    return *this;
 }
 
-inline DynamicData DynamicData::operator + (const ReadableDynamicDataRef& other) const
+inline bool DynamicData::operator ! () const
 {
     switch(type_.kind())
     {
+        case TypeKind::CHAR_8_TYPE:
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(char);
+        case TypeKind::CHAR_16_TYPE:
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(char16_t);
+        case TypeKind::BOOLEAN_TYPE:
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(bool);
         case TypeKind::INT_8_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(int8_t, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(int8_t);
         case TypeKind::UINT_8_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(uint8_t, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(uint8_t);
         case TypeKind::INT_16_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(int16_t, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(int16_t);
         case TypeKind::UINT_16_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(uint16_t, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(uint16_t);
         case TypeKind::INT_32_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(int32_t, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(int32_t);
         case TypeKind::UINT_32_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(uint32_t, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(uint32_t);
         case TypeKind::INT_64_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(int64_t, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(int64_t);
         case TypeKind::UINT_64_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(uint64_t, +);
-        case TypeKind::FLOAT_32_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(float, +);
-        case TypeKind::FLOAT_64_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(double, +);
-        case TypeKind::FLOAT_128_TYPE:
-            DYNAMIC_DATA_OPERATOR_RESULT(long double, +);
+            DYNAMIC_DATA_NOT_OPERATOR_RESULT(uint64_t);
+        case TypeKind::STRING_TYPE:
+            return this->value<std::string>().empty();
+        case TypeKind::WSTRING_TYPE:
+            return this->value<std::wstring>().empty();
         default:
-            xtypes_assert(false,
-                "operator<<() isn't available for types '" << type_.name() << "' and '" << other.type().name() << "'.");
-            return *this;
+            xtypes_assert(false, "operator!() isn't available for type '" << "'.");
     }
+}
 
+#define DYNAMIC_DATA_LOGIC_OPERATION(TYPE, OPERATOR) \
+{\
+    switch(other.type().kind())\
+    {\
+        case TypeKind::CHAR_8_TYPE:\
+            return (this->ReadableDynamicDataRef::value<TYPE>() OPERATOR other.ReadableDynamicDataRef::value<char>());\
```

Comment:
Should be equivalent, am I wrong?
```suggestion
            return (this->value<TYPE>() OPERATOR other.value<char>());\
```